lib/kube/grpc/grpc.go

```diff
@@ -0,0 +1,314 @@
+	}
+	if c.Emitter == nil {
+		return trace.BadParameter("missing parameter Emitter")
+	}
+	if c.Context == nil {
+		c.Context = context.TODO()
+	}
+	if c.KubeProxyAddr == "" {
+		return trace.BadParameter("missing parameter KubeProxyAddr")
+	}
+	if c.ClusterName == "" {
+		return trace.BadParameter("missing parameter ClusterName")
+	}
+	if c.Component == "" {
+		c.Component = "kube.grpc"
+	}
+	if c.Log == nil {
+		c.Log = logrus.New()
+	}
+	c.Log = c.Log.WithFields(logrus.Fields{trace.Component: c.Component})
+	return nil
+}
+
+// ListKubernetesResources returns the list of pods available for the user for
+// the specified Kubernetes cluster and namespace.
+func (s *Server) ListKubernetesResources(ctx context.Context, req *proto.ListKubernetesResourcesRequest) (*proto.ListKubernetesResourcesResponse, error) {
+	userContext, err := s.authorize(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	if req.UseSearchAsRoles || req.UsePreviewAsRoles {
+		var extraRoles []string
+		if req.UseSearchAsRoles {
+			extraRoles = append(extraRoles, userContext.Checker.GetAllowedSearchAsRoles()...)
+		}
+		if req.UsePreviewAsRoles {
+			extraRoles = append(extraRoles, userContext.Checker.GetAllowedPreviewAsRoles()...)
+		}
+		clusterName, err := s.cfg.AuthClient.GetClusterName()
+		if err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+		if err := userContext.UseExtraRoles(s.cfg.AuthClient, clusterName.GetClusterName(), extraRoles); err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+		if err := s.emitAuditEvent(ctx, userContext, req); err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+	}
+	// We use the unmapped identity here because Kube Proxy will handle
+	// the forwarding of the request to the correct leaf cluster if that's the case
+	// and it handles the mapping of the identity to the leaf cluster.
+	identity := userContext.UnmappedIdentity.GetIdentity()
+	identity.KubernetesCluster = req.KubernetesCluster
+	identity.Groups = userContext.Checker.RoleNames()
+	identity.RouteToCluster = req.TeleportCluster
+	switch {
+	case req.ResourceType == types.KindKubePod:
+		rsp, err := s.listKubernetesPods(ctx, identity, true, req)
+		return rsp, trail.ToGRPC(err)
+	default:
+		return nil, trail.ToGRPC(trace.BadParameter("unsupported resource type %q", req.ResourceType))
+	}
+}
+
+// authorize checks if the user is authorized to connect to the cluster.
+func (s *Server) authorize(ctx context.Context) (*auth.Context, error) {
+	accessDeniedMsg := "access denied"
+	userContext, err := s.cfg.Authz.Authorize(ctx)
+
+	switch {
+	case err == nil:
+		return userContext, nil
+	// propagate connection problem error so we can differentiate
+	// between connection failed and access denied
+	case trace.IsConnectionProblem(err):
+		return nil, trace.ConnectionProblem(err, "[07] failed to connect to the database")
+	case trace.IsAccessDenied(err):
+		// don't print stack trace, just log the warning
+		s.cfg.Log.Warn(err)
```

Comment:
Should we add any additional context to this warning ?

Reply:
This is just to avoid printing stack traces to access denied requests. 
If running in debug mode we already print the reason for the access denied internally and we don't need to print the stack trace for those cases.